pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java

```diff
@@ -0,0 +1,293 @@
+    }
+
+    // read ledger entries.
+    private int readEntries() throws IOException {
+        checkState(bytesReadOffset >= dataBlockHeaderLength);
+        checkState(bytesReadOffset < dataBlockFullOffset);
+        checkState(ledger != null);
+
+        try {
+            // once reach the end of entry buffer, start a new read.
+            if (bytesReadOffset == blockBytesHave) {
+                readLedgerEntriesOnce();
+                if (log.isDebugEnabled()) {
+                    log.debug("After readLedgerEntriesOnce: bytesReadOffset: {}, blockBytesHave: {}",
+                        bytesReadOffset, blockBytesHave);
+                }
+            }
+
+            // read entry header from entriesHeadBuf, or entry content from entriesByteBuf.
+            int entriesHeadIndex = entriesHeadBuf.readerIndex();
+            if (entriesHeadIndex % entryHeaderSize == 0) {
+                if (entriesHeadIndex / entryHeaderSize > 0) {
+                    if (entryByteBuf.readableBytes() == 0) {
+                        entryByteBuf.release();
+                        // get entry content of this entry header.
+                        if (entriesHeadIndex / entryHeaderSize < entriesByteBuf.size()) {
+                            entryByteBuf = entriesByteBuf.get(entriesHeadIndex / entryHeaderSize);
+                        }
+                        bytesReadOffset ++;
+                        return entriesHeadBuf.readByte();
+                    }
+
+                    if (entryByteBuf.readableBytes() > 0) {
+                        bytesReadOffset ++;
+                        return entryByteBuf.readByte();
+                    } else {
+                        // should never reach here.
+                        throw new IOException("ByteBuf readableBytes < 0.  value: " + entryByteBuf.readableBytes());
+                    }
+                } else {
+                    bytesReadOffset ++;
+                    return entriesHeadBuf.readByte();
+                }
+            } else {
+                bytesReadOffset ++;
+                return entriesHeadBuf.readByte();
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Exception when get CompletableFuture<LedgerEntries>. ", e);
+            throw new IOException(e);
+        }
+    }
+
+    // read entries from ledger, and pre-handle the returned ledgerEntries.
+    private void readLedgerEntriesOnce() throws InterruptedException, ExecutionException {
+        checkState(bytesReadOffset == blockBytesHave);
+
+        // clear data and pointer of former read
+        if (!entriesByteBuf.isEmpty()) {
+            entriesByteBuf.forEach(byteBuf -> byteBuf.release());
+            entriesByteBuf.clear();
+        }
+        if (entriesHeadBuf != null) {
+            entriesHeadBuf.release();
+            entriesHeadBuf = null;
+        }
+        entriesHeadBuf = Unpooled.wrappedBuffer(entriesHead);
```

Comment:
As I said above, you can get rid of this buffer completely if you use a CompositeBuffer.

```
ByteBuf entryLen = Unpooled.buffer(4);
entryLen.writeInt(/* etc */);
entriesByteBuf.add(Unpooled.wrappedBuffer(entryLen, entryPayload));
```